crates/iceberg-catalog/src/service/task_queue/tabular_expiration_queue.rs

```diff
@@ -0,0 +1,334 @@
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to get warehouse: {:?}", e);
+            e
+        })?;
+
+    let secret = maybe_get_secret(warehouse.storage_secret_id, secret_state)
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to get secret: {:?}", e);
+            e
+        })?;
+
+    let fio = warehouse
+        .storage_profile
+        .file_io(secret.as_ref())
+        .map_err(|e| {
+            tracing::error!("Failed to get storage profile: {:?}", e);
+            e
+        })?;
+    tracing::debug!("Got FileIO");
+
+    handle_old_metadata(
+        warehouse_ident,
+        &fio,
+        &table_metadata.table_metadata,
+        delete_queue,
+        task_id,
+    )
+    .await
+    .map_err(|e| {
+        tracing::error!("Failed to handle old metadata: {:?}", e);
+        e
+    })?;
+
+    let location = table_metadata.metadata_location.as_deref().ok_or_else(|| {
+        let err = ErrorModel::internal(
+            "Table metadata location not found",
+            "MetadataLocationNotFound",
+            None,
+        );
+        tracing::error!("Table metadata location not found: {:?}", err);
+        err
+    })?;
+
+    find_files(
+        &table_metadata.table_metadata,
+        location,
+        warehouse_ident,
+        delete_queue,
+        &fio,
+        task_id,
+    )
+    .await
+    .map_err(|e| {
+        tracing::error!("Failed to find files: {:?}", e);
+        e
+    })?;
+
+    C::drop_table(
+        tabular_id.into(),
+        DropFlags {
+            hard_delete: true,
+            purge: true,
+        },
+        trx.transaction(),
+    )
+    .await
+    .map_err(|e| {
+        tracing::error!("Failed to drop table: {:?}", e);
+        e
+    })?;
+
+    trx.commit().await.map_err(|e| {
+        tracing::error!("Failed to commit transaction: {:?}", e);
+        e
+    })?;
+
+    Ok(())
+}
+
+async fn handle_old_metadata(
+    warehouse_id: WarehouseIdent,
+    fio: &FileIO,
+    meta: &TableMetadata,
+    fetcher: &impl TaskQueue<Task = Deletion, Input = DeleteInput>,
+    task_id: Uuid,
+) -> Result<()> {
+    for old_meta in &meta.metadata_log {
+        // old metadata doesnt exist in db hence we're going to s3
+        let input = fio
+            .new_input(old_meta.metadata_file.as_str())
+            .map_err(|e| {
+                ErrorModel::internal("Failed to create input", "FileIOError", Some(Box::new(e)))
+            })?;
+        let metadata = input.read().await.map_err(|e| {
+            ErrorModel::internal("Failed to read input", "FileIOError", Some(Box::new(e)))
+        })?;
+        let meta = serde_json::from_slice::<TableMetadata>(&metadata).map_err(|e| {
+            ErrorModel::internal(
+                "Failed to deserialize metadata",
+                "SerdeJsonError",
+                Some(Box::new(e)),
+            )
+        })?;
+
+        find_files::<_>(
```

Comment:
I don't think this is needed.
Snapshots are copied from one metadata version to the next unless they are explicitly removed by "expire_snapshots" procedure. The expire snapshot procedure should do the walk and expire those.